postprocessing/science/spectral_corner_frequency_from_surface_xdmf.py

```diff
@@ -0,0 +1,384 @@
+
+parser.add_argument('--avgSWaveVelocity', nargs=1, metavar=('avgSWaveVelocity'),
+                    help='provide an average S-wave velocity near the surface (otherwise it is approximated from P-arrivals)',
+                    type=float)
+args = parser.parse_args()
+
+def LoadSingleComponentParallel(component, surfaceData=True):
+    chunks = np.int_(np.linspace(0, nElements, nprocs+1)) if surfaceData else np.int_(np.linspace(0, nElementsFault, nprocs+1))
+    chunks = np.array([chunks[:-1], chunks[1:]-chunks[:-1], np.full(nprocs, component)]).T
+    chunks = [tuple(i) for i in chunks]
+    
+    if surfaceData:
+        p = Pool(nprocs)
+        result = p.map(ReadSurfaceWithPool, chunks)
+        p.close()
+        p.join()
+        return np.concatenate(result, axis=1).T[::stepsize]
+    else:
+        p = Pool(nprocs)
+        result = p.map(ReadFaultWithPool, chunks)
+        p.close()
+        p.join()
+        return np.concatenate(result, axis=1).T[::stepsize]
+    
+def ReadSurfaceWithPool(chunk):
+    return surfacexdmf.ReadDataChunk(chunk[2], int(chunk[0]), int(chunk[1]))[timeIndices[0]:timeIndices[1]]
+
+def ReadFaultWithPool(chunk):
+    return faultxdmf.ReadDataChunk(chunk[2], int(chunk[0]), int(chunk[1]))[timeIndicesFault[0]:timeIndicesFault[1]]
+
+def ComputeTriangleMidpoints(geom, connect):
+    """Generates an array with coordinates of triangle midpoints (same dimension as connect)"""
+    xyz = np.zeros_like(connect, dtype=float)
+    xyz = (1./3.)*(geom[connect[:,0],:]+geom[connect[:,1],:]+geom[connect[:,2],:])   
+    return xyz
+
+def CalculateSlipCentroid(faultxdmf, events=0):
+    if events == 0:
+        ASl = faultxdmf.ReadData("ASl", idt=lastTimeStep).T
+    elif events == 1:
+        ASl = faultxdmf.ReadData("ASl", idt=int(lastTimeStep/2)).T   
+    else:
+        ASl = faultxdmf.ReadData("ASl", idt=lastTimeStep).T - faultxdmf.ReadData("ASl", idt=int(lastTimeStep/2)).T 
+    return np.average(faultxyz, axis=0, weights=ASl)
+
+def ComputeBackazimuth(xyz, centroid):
+    ba = np.arctan2(xyz[:,0]-centroid[0], xyz[:,1]-centroid[1]) % (2*np.pi)
+    ba = np.where(ba < np.pi, ba+np.pi, ba-np.pi)
+    return ba
+
+def RotateToRadialTransversal(u, v, xyz, centroid):
+    ba = ComputeBackazimuth(xyz, centroid)
+    ba = ba.reshape(-1,1)
+    r = -u * np.sin(ba) - v * np.cos(ba)
+    t = -u * np.cos(ba) + v * np.sin(ba)
+    return r, t
+
+def ComputeCornerFrequencyParallel(component, pphases, windowduration):
+    chunk_inds = np.int_(np.linspace(0, component.shape[0], nprocs+1))
+    chunks = []
+    for i in range(nprocs):
+        chunks.append([component[chunk_inds[i]:chunk_inds[i+1]], pphases[chunk_inds[i]:chunk_inds[i+1]], 
+                     windowduration[chunk_inds[i]:chunk_inds[i+1]]])
+    chunks = [tuple(i) for i in chunks]  
+       
+    p = Pool(nprocs)
+    result = p.map(ComputeCornerFrequencyArray, chunks)
```

Comment:
I think your use of pool.map is unnecessarily complicated.
Then you do not need to divide your data in chunks.
Why don't you just use pool.map on ComputeSingleCornerFrequency?

Reply:
ok I guess it is because you want to leverage on broadcasting.
But using pool.map on ComputeSingleCornerFrequency would allow to scale to higher CPU counts. Anyway, your choice.

Reply:
I changed the functions to avoid "ComputeCornerFrequencyArray"